From 0f5818070fd835afc51a1bdb56ad082b24fcc6aa Mon Sep 17 00:00:00 2001
From: Hybridmax <hybridmax95@gmail.com>
Date: Sun, 8 Mar 2015 18:45:11 +0100
Subject: [PATCH] GPU PWR LEVELS: Fixed super old bug in gpu thermal pwr level

---
 drivers/gpu/msm/kgsl_pwrctrl.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/msm/kgsl_pwrctrl.c b/drivers/gpu/msm/kgsl_pwrctrl.c
index 87b3e78..c9ade18 100755
--- a/drivers/gpu/msm/kgsl_pwrctrl.c
+++ b/drivers/gpu/msm/kgsl_pwrctrl.c
@@ -397,7 +397,15 @@ static int kgsl_pwrctrl_max_gpuclk_store(struct device *dev,
 	if (level < 0)
 		goto done;
 
-	pwr->thermal_pwrlevel = level;
+	if (level > 7)
+		level = 7;
+
+	if (level == 0)
+		pwr->thermal_pwrlevel = level;
+	else if (level == 1)
+		pwr->thermal_pwrlevel = level;
+	else
+		pwr->thermal_pwrlevel = level - 2;
 
 	/*
 	 * if the thermal limit is lower than the current setting,
