From 0580fcf298a3e021d88adcef500ab3ce18290266 Mon Sep 17 00:00:00 2001
From: ausdim <koronaios@gmail.com>
Date: Wed, 5 Mar 2014 23:16:42 +0200
Subject: [PATCH] logger: Add an enable toggle for entry storage (Thanks to
 AndreiLux)

---
 drivers/staging/android/logger.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/staging/android/logger.c b/drivers/staging/android/logger.c
index 2700382..8fb6455 100755
--- a/drivers/staging/android/logger.c
+++ b/drivers/staging/android/logger.c
@@ -33,6 +33,10 @@
 static char klog_buf[256];
 #endif
 
+static unsigned int log_enabled = 1;
+
+module_param(log_enabled, uint, S_IWUSR | S_IRUGO);
+
 /*
  * struct logger_log - represents a specific log, such as 'main' or 'radio'
  *
@@ -466,6 +470,9 @@ ssize_t logger_aio_write(struct kiocb *iocb, const struct iovec *iov,
 	struct timespec now;
 	ssize_t ret = 0;
 
+  	if (!log_enabled)
+     	return 0;
+
 	now = current_kernel_time();
 
 	header.pid = current->tgid;
